File: presentation/src/main/java/com/github/anrimian/musicplayer/di/app/AppModule.java
package com.github.anrimian.musicplayer.di.app;

import android.content.Context;

import androidx.annotation.NonNull;

import com.github.anrimian.musicplayer.data.repositories.logger.LoggerRepositoryImpl;
import com.github.anrimian.musicplayer.domain.controllers.SystemServiceController;
import com.github.anrimian.musicplayer.domain.interactors.analytics.Analytics;
import com.github.anrimian.musicplayer.domain.interactors.player.LibraryPlayerInteractor;
import com.github.anrimian.musicplayer.domain.interactors.settings.DisplaySettingsInteractor;
import com.github.anrimian.musicplayer.domain.repositories.LoggerRepository;
import com.github.anrimian.musicplayer.infrastructure.analytics.AnalyticsImpl;
import com.github.anrimian.musicplayer.infrastructure.service.SystemServiceControllerImpl;
import com.github.anrimian.musicplayer.ui.common.images.CoverImageLoader;
import com.github.anrimian.musicplayer.ui.common.theme.ThemeController;
import com.github.anrimian.musicplayer.ui.notifications.NotificationsDisplayer;
import com.github.anrimian.musicplayer.ui.notifications.builder.AppNotificationBuilder;
import com.github.anrimian.musicplayer.ui.widgets.WidgetUpdater;
import com.github.anrimian.musicplayer.utils.logger.AppLogger;
import com.github.anrimian.musicplayer.utils.logger.FileLog;

import javax.annotation.Nonnull;
import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import io.reactivex.rxjava3.core.Scheduler;

import static com.github.anrimian.musicplayer.di.app.SchedulerModule.UI_SCHEDULER;

/**
 * Created on 11.02.2017.
 */

@Module
public class AppModule {

    private Context appContext;

    public AppModule(@NonNull Context appContext) {
        this.appContext = appContext;
    }

    @Provides
    @NonNull
    Context provideAppContext() {
        return appContext;
    }

    @Provides
    @Nonnull
    @Singleton
    NotificationsDisplayer provideNotificationsController(Context context,
                                                          AppNotificationBuilder notificationBuilder,
                                                          CoverImageLoader coverImageLoader) {
        return new NotificationsDisplayer(context, notificationBuilder, coverImageLoader);
    }

    @Provides
    @Nonnull
    @Singleton
    AppNotificationBuilder appNotificationBuilder() {
        return new AppNotificationBuilder();
    }

    @Provides
    @Nonnull
    @Singleton
    SystemServiceController systemServiceController(Context context) {
        return new SystemServiceControllerImpl(context);
    }

    @Provides
    @Nonnull
    @Singleton
    Analytics analytics(FileLog fileLog) {
        return new AnalyticsImpl(fileLog);
    }

    @Provides
    @Nonnull
    @Singleton
    FileLog fileLog(Context context) {
        return new FileLog(context);
    }

    @Provides
    @Nonnull
    @Singleton
    LoggerRepository loggerRepository(Context context) {
        return new LoggerRepositoryImpl(context);
    }

    @Provides
    @Nonnull
    @Singleton
    AppLogger appLogger(FileLog fileLog, LoggerRepository loggerRepository) {
        return new AppLogger(fileLog, loggerRepository);
    }

    @Provides
    @NonNull
    @Singleton
    WidgetUpdater widgetUpdater(Context context,
                                LibraryPlayerInteractor playerInteractor,
                                DisplaySettingsInteractor displaySettingsInteractor,
                                @Named(UI_SCHEDULER) Scheduler scheduler) {
        return new WidgetUpdater(context, playerInteractor, displaySettingsInteractor, scheduler);
    }

    @Provides
    @NonNull
    @Singleton
    ThemeController themeController(Context context) {
        return new ThemeController(context);
    }
}

File: app-wear/src/main/java/com/github/anrimian/musicplayer/wear/di/app/AppModule.kt
package com.github.anrimian.musicplayer.wear.di.app

import android.content.Context
import com.github.anrimian.musicplayer.di.app.SchedulerModule.Companion.IO_SCHEDULER
import com.github.anrimian.musicplayer.di.app.SchedulerModule.Companion.UI_SCHEDULER
import com.github.anrimian.musicplayer.domain.interactors.analytics.Analytics
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.wear.Constants
import com.github.anrimian.musicplayer.wear.data.WearStateRepository
import com.github.anrimian.musicplayer.wear.data.repositories.HostDeviceRepository
import com.github.anrimian.musicplayer.wear.domain.WearStateInteractor
import com.github.anrimian.musicplayer.wear.domain.controllers.RemoteStateController
import com.github.anrimian.musicplayer.wear.domain.queue.PlayQueueInteractor
import com.github.anrimian.musicplayer.wear.infrastructure.DeviceConnectionController
import com.github.anrimian.musicplayer.wear.infrastructure.analytics.WearableAnalytics
import com.github.anrimian.musicplayer.wear.infrastructure.remote.RemoteStateControllerImpl
import com.github.anrimian.musicplayer.wear.ui.MainPresenter
import com.github.anrimian.musicplayer.wear.ui.common.error.WearErrorParser
import com.github.anrimian.musicplayer.wear.utils.logger.WearableFileLogger
import dagger.Module
import dagger.Provides
import io.reactivex.rxjava3.core.Scheduler
import javax.inject.Named
import javax.inject.Singleton

@Module
class AppModule(private val context: Context) {

    @Provides
    fun context() = context

    @Provides
    @Singleton
    fun wearStateRepository(context: Context) = WearStateRepository(context)

    @Provides
    @Singleton
    fun remoteStateController(
        context: Context
    ): RemoteStateController = RemoteStateControllerImpl(context)

    @Provides
    @Singleton
    fun wearStateInteractor(
        wearStateRepository: WearStateRepository,
        deviceRepository: HostDeviceRepository,
        remoteStateController: RemoteStateController,
    ) = WearStateInteractor(
        wearStateRepository,
        deviceRepository,
        remoteStateController
    )

    @Provides
    @Singleton
    fun playQueueInteractor(
        hostDeviceRepository: HostDeviceRepository
    ) = PlayQueueInteractor(
        hostDeviceRepository
    )

    @Provides
    @Singleton
    fun hostDeviceRepository(
        deviceConnectionController: DeviceConnectionController,
        @Named(IO_SCHEDULER) ioScheduler: Scheduler,
    ) = HostDeviceRepository(
        deviceConnectionController,
        ioScheduler,
        Constants.EVENT_TIMEOUT_MILLIS
    )

    @Provides
    fun mainPresenter(
        wearStateInteractor: WearStateInteractor,
        playQueueInteractor: PlayQueueInteractor,
        @Named(UI_SCHEDULER) uiScheduler: Scheduler,
        errorParser: ErrorParser,
    ) = MainPresenter(
        wearStateInteractor,
        playQueueInteractor,
        uiScheduler,
        errorParser
    )

    @Provides
    @Singleton
    fun errorParser(
        context: Context,
        analytics: Analytics,
    ): ErrorParser = WearErrorParser(context, analytics)

    @Provides
    @Singleton
    fun deviceConnectionController(
        context: Context,
        @Named(IO_SCHEDULER) ioScheduler: Scheduler,
    ) = DeviceConnectionController(context, ioScheduler)

    @Provides
    @Singleton
    fun analytics(fileLog: WearableFileLogger): Analytics {
        return WearableAnalytics(fileLog)
    }

    @Provides
    @Singleton
    fun fileLog(): WearableFileLogger {
        return WearableFileLogger()
    }

}